File: runtime/kernel/kernel_bridge.R
#!/usr/bin/env Rscript
# Minimal local R kernel for the ZeroWall Science notebook.
#
# A persistent process that holds one environment across cells (shared state,
# like a Jupyter kernel). The host writes the cell's code to a file (argv[1]),
# then sends one line "<id>" on stdin; we evaluate that file in the global
# environment and write ONE JSON response line back to stdout:
#
#     response: {"id","ok","stdout","result","error","image","truncated"}\n
#
# Base R only — no jsonlite/IRkernel — so it runs against any R install,
# offline, with no model key. `result` mirrors Jupyter: the printed value of the
# cell's final expression when it is visible, else null. `image` is a base64 PNG
# when the cell drew a plot.
#
# `sink()` only captures R's own connections, so a cell calling system() hands
# the child the process's real stdout — the protocol stream. Those two calls are
# therefore shadowed for the duration of each cell and their output folded into
# the cell's own, so a shell command can never desynchronize the host.

args <- commandArgs(trailingOnly = TRUE)
codefile <- args[1]
options(warn = 1) # surface warnings inline (stdout), not deferred to session end

# Cap on one cell's captured output; matches the Python bridge.
MAX_OUTPUT_CHARS <- 200000L
MAX_IMAGE_BYTES <- 4000000L

# JSON-escape a scalar string. fixed = TRUE keeps replacements literal so we do
# not fight regex metacharacters; control chars left after \n\r\t are dropped.
json_escape <- function(s) {
  if (length(s) != 1) s <- paste(s, collapse = "\n")
  if (is.na(s)) return("")
  s <- gsub("\\", "\\\\", s, fixed = TRUE)
  s <- gsub("\"", "\\\"", s, fixed = TRUE)
  s <- gsub("\n", "\\n", s, fixed = TRUE)
  s <- gsub("\r", "\\r", s, fixed = TRUE)
  s <- gsub("\t", "\\t", s, fixed = TRUE)
  s <- gsub("[[:cntrl:]]", "", s)
  s
}

emit <- function(id, ok, out, result, error, image, truncated) {
  parts <- c(
    paste0("\"id\":\"", json_escape(id), "\""),
    paste0("\"ok\":", if (ok) "true" else "false"),
    paste0("\"stdout\":\"", json_escape(out), "\""),
    if (is.null(result)) "\"result\":null" else paste0("\"result\":\"", json_escape(result), "\""),
    if (is.null(error)) "\"error\":null" else paste0("\"error\":\"", json_escape(error), "\""),
    if (is.null(image)) "\"image\":null" else paste0("\"image\":\"", image, "\""),
    paste0("\"truncated\":", if (truncated) "true" else "false")
  )
  cat(paste0("{", paste(parts, collapse = ","), "}"), "\n", sep = "")
  flush(stdout())
}

# Clip captured output, keeping the head and the tail.
truncate_output <- function(text) {
  if (nchar(text) <= MAX_OUTPUT_CHARS) return(list(text = text, truncated = FALSE))
  keep <- MAX_OUTPUT_CHARS %/% 2L
  dropped <- nchar(text) - 2L * keep
  list(
    text = paste0(substr(text, 1L, keep),
                  "\n... [", dropped, " characters omitted] ...\n",
                  substr(text, nchar(text) - keep + 1L, nchar(text))),
    truncated = TRUE
  )
}

# Base64-encode raw bytes. Base R ships no base64 encoder (tools::base64encode
# arrived late and is absent from common installs, e.g. 4.3.x), and this file
# takes no package dependencies, so encode it directly: 3 bytes -> 4 chars,
# with the standard "=" tail padding.
base64_raw <- function(raw) {
  alphabet <- c(LETTERS, letters, 0:9, "+", "/")
  n <- length(raw)
  pad <- (3L - n %% 3L) %% 3L
  if (pad > 0L) raw <- c(raw, as.raw(rep(0L, pad)))
  v <- as.integer(raw)
  triples <- matrix(v, nrow = 3L)
  bits <- triples[1L, ] * 65536L + triples[2L, ] * 256L + triples[3L, ]
  chars <- rbind(
    alphabet[bits %/% 262144L + 1L],
    alphabet[(bits %/% 4096L) %% 64L + 1L],
    alphabet[(bits %/% 64L) %% 64L + 1L],
    alphabet[bits %% 64L + 1L]
  )
  out <- paste(chars, collapse = "")
  if (pad > 0L) {
    out <- paste0(substr(out, 1L, nchar(out) - pad), strrep("=", pad))
  }
  out
}

# Base64 PNG of whatever the cell drew, or NULL. Plotting goes to an offscreen
# PNG device opened per cell, so a plot() call produces a figure instead of
# silently going nowhere in a headless process.
capture_plot_png <- function(path) {
  if (!file.exists(path)) return(NULL)
  size <- file.info(path)$size
  if (is.na(size) || size == 0 || size > MAX_IMAGE_BYTES) return(NULL)
  tryCatch(base64_raw(readBin(path, "raw", n = size)), error = function(e) NULL)
}

run_cell <- function(code, plotfile) {
  exprs <- tryCatch(parse(text = code), error = function(e) e)
  if (inherits(exprs, "error")) {
    return(list(ok = FALSE, stdout = "", result = NULL,
                error = paste0("Error: ", conditionMessage(exprs)),
                image = NULL, truncated = FALSE))
  }
  captured <- character(0)
  buf <- textConnection("captured", open = "w", local = TRUE)

  # An offscreen device, so plot() draws a figure instead of failing or going
  # nowhere in a headless process. Absent graphics support is not an error:
  # the cell still runs, it just produces no image.
  unlink(plotfile)
  has_device <- tryCatch({
    grDevices::png(filename = plotfile, width = 800, height = 600, res = 110)
    TRUE
  }, error = function(e) FALSE, warning = function(w) FALSE)
  # An untouched device still writes a blank PNG on close. `dev.size` cannot
  # tell drawn from empty, so ask the display list: an empty one means the
  # cell drew nothing and the figure must be dropped, not shown as a white box.
  if (has_device) tryCatch(grDevices::dev.control("enable"), error = function(e) NULL)

  # system()/system2() bypass sink(): they hand the child the kernel's own
  # descriptors, so the child's output would arrive as a bare line between
  # responses and desynchronize the host. Shadow both for the duration of the
  # cell, routing the child through files.
  #
  # `intern = TRUE` is NOT usable here: on Windows it closes the kernel's
  # inherited stdin, so the read loop sees EOF and the kernel exits — a shell
  # command would silently end the session. Explicit file redirection with a
  # private stdin file leaves our descriptors untouched.
  run_child <- function(command, args) {
    outfile <- tempfile()
    infile <- tempfile()
    file.create(infile)
    on.exit(unlink(c(outfile, infile)), add = TRUE)
    status <- tryCatch(
      base::system2(command, args, stdout = outfile, stderr = outfile, stdin = infile),
      error = function(e) {
        cat("Error: ", conditionMessage(e), "\n", sep = "")
        127L
      }
    )
    if (file.exists(outfile)) {
      text <- readLines(outfile, warn = FALSE)
      if (length(text)) cat(paste(text, collapse = "\n"), "\n", sep = "")
    }
    invisible(if (is.numeric(status)) as.integer(status) else 0L)
  }
  # base::system() takes one command line; hand it to the shell so pipes and
  # redirection keep working.
  shell_cmd <- if (.Platform$OS.type == "windows") "cmd" else "/bin/sh"
  shell_flag <- if (.Platform$OS.type == "windows") "/c" else "-c"
  assign("system", function(command, ...) run_child(shell_cmd, c(shell_flag, shQuote(command))),
         envir = globalenv())
  assign("system2", function(command, args = character(), ...) run_child(command, args),
         envir = globalenv())

  sink(buf)
  sink(buf, type = "message")
  result <- NULL
  err <- NULL
  tryCatch({
    n <- length(exprs)
    if (n > 0) for (i in seq_len(n)) {
      wv <- withVisible(eval(exprs[[i]], envir = globalenv()))
      if (wv$visible) {
        printed <- paste(utils::capture.output(print(wv$value)), collapse = "\n")
        if (i == n) result <- printed else cat(printed, "\n", sep = "")
      }
    }
  }, error = function(e) {
    err <<- paste0("Error: ", conditionMessage(e))
  })
  sink(type = "message")
  sink()
  close(buf)
  # Ask the display list whether the cell actually drew: an untouched device
  # still writes a blank PNG, and a white box is not a figure.
  drew <- FALSE
  if (has_device) {
    drew <- tryCatch(length(grDevices::recordPlot()[[1]]) > 0L, error = function(e) FALSE)
    tryCatch(grDevices::dev.off(), error = function(e) NULL)
  }
  # Restore the real system()/system2() so a later cell defining its own is
  # not fighting a leftover shim.
  suppressWarnings(rm(list = c("system", "system2"), envir = globalenv()))

  image <- if (drew) capture_plot_png(plotfile) else NULL
  unlink(plotfile)
  clipped <- truncate_output(paste(captured, collapse = "\n"))
  list(ok = is.null(err), stdout = clipped$text, result = result, error = err,
       image = image, truncated = clipped$truncated)
}

con <- file("stdin", open = "r")
# Per-kernel scratch files, named off the code file so concurrent notebooks
# never share them.
plotfile <- paste0(codefile, ".png")
repeat {
  line <- readLines(con, n = 1)
  if (length(line) == 0) break # host closed stdin -> exit
  id <- trimws(line)
  if (nchar(id) == 0) next
  code <- tryCatch(paste(readLines(codefile, warn = FALSE), collapse = "\n"),
                   error = function(e) "")
  r <- run_cell(code, plotfile)
  emit(id, r$ok, r$stdout, r$result, r$error, r$image, r$truncated)
}

File: runtime/kernel/kernel_bridge.py
#!/usr/bin/env python3
"""Minimal local Python kernel for the ZeroWall Science notebook.

A persistent process that holds one namespace across cells (shared state, like a
Jupyter kernel) and speaks a line-delimited JSON protocol over stdin/stdout:

    request : {"id": "<str>", "code": "<str>"}\\n
    response: {"id","ok","stdout","result","error","image","truncated"}\\n

Standard library only — no ipykernel/ZMQ — so it runs against whatever Python the
user has, offline, with no model key. `result` mirrors Jupyter: the repr of the
final expression when a cell ends in one, else null. `image` is a base64 PNG when
the cell left a matplotlib figure open.

The protocol stream is PRIVATE. Cell output cannot reach it: fd 1 is duplicated
to a reserved descriptor at startup and every response is written there, while
fd 1 itself is redirected into the cell's capture buffer. A cell that spawns a
subprocess, or writes to fd 1 directly, therefore produces cell output — not a
corrupt response line that would desynchronize the host and cost the user their
whole session state.
"""
import ast
import base64
import io
import json
import os
import sys
import tempfile
import traceback
from contextlib import redirect_stderr, redirect_stdout

# Cap on a single cell's captured output. A runaway loop otherwise builds one
# unbounded JSON line (a `for i in range(200000): print(i)` cell measured
# 1.3 MB), which the host must buffer whole before the user sees anything.
MAX_OUTPUT_CHARS = 200_000

# Cap on an emitted figure. Beyond this the PNG is dropped rather than sent —
# the response is one line, and a huge image would stall the notebook.
MAX_IMAGE_BYTES = 4_000_000


def _truncate(text: str) -> tuple[str, bool]:
    """Clip captured output to MAX_OUTPUT_CHARS, keeping the head and tail."""
    if len(text) <= MAX_OUTPUT_CHARS:
        return text, False
    keep = MAX_OUTPUT_CHARS // 2
    dropped = len(text) - 2 * keep
    return f"{text[:keep]}\n... [{dropped} characters omitted] ...\n{text[-keep:]}", True


def capture_figure() -> str | None:
    """Base64 PNG of the current matplotlib figure, if a cell left one open.

    Matplotlib is optional and often absent; importing it must never be an
    error. Only an already-imported module is inspected — importing it here
    would cost seconds on the first cell of every session, and a user who
    never plots should not pay that.
    """
    plt = sys.modules.get("matplotlib.pyplot")
    if plt is None:
        return None
    try:
        if not plt.get_fignums():
            return None
        buf = io.BytesIO()
        plt.savefig(buf, format="png", bbox_inches="tight", dpi=110)
        plt.close("all")  # Jupyter semantics: a shown figure is consumed
        raw = buf.getvalue()
        if len(raw) > MAX_IMAGE_BYTES:
            return None
        return base64.b64encode(raw).decode("ascii")
    except Exception:
        # A broken backend must not fail the cell — the user's code ran fine.
        return None


def run_cell(ns: dict, code: str):
    """Execute `code` in namespace `ns`.

    Returns (stdout, result_repr_or_None, error_or_None, image_or_None,
    truncated). Output written to fd 1 by a subprocess is captured too, via
    the temporary file fd 1 points at for the duration of the cell.
    """
    out = io.StringIO()
    try:
        parsed = ast.parse(code, mode="exec")
    except SyntaxError:
        return "", None, traceback.format_exc(limit=1), None, False

    body = parsed.body
    result = None
    # Jupyter behaviour: if the cell ends in an expression, show its value.
    tail_expr = None
    if body and isinstance(body[-1], ast.Expr):
        last = body.pop()
        assert isinstance(last, ast.Expr)
        tail_expr = ast.Expression(last.value)

    error = None
    with tempfile.TemporaryFile(mode="w+b") as sink:
        # Point the real fd 1 and fd 2 at the sink for the duration of the
        # cell, so a subprocess (which inherits descriptors, not sys.stdout)
        # writes into the cell's output instead of into the protocol stream.
        saved = os.dup(1), os.dup(2)
        try:
            os.dup2(sink.fileno(), 1)
            os.dup2(sink.fileno(), 2)
            try:
                with redirect_stdout(out), redirect_stderr(out):
                    if body:
                        exec(compile(ast.Module(body, []), "<cell>", "exec"), ns)  # noqa: S102
                    if tail_expr is not None:
                        value = eval(compile(tail_expr, "<cell>", "eval"), ns)  # noqa: S307
                        if value is not None:
                            result = repr(value)
            except BaseException:  # noqa: BLE001 - report, never kill the kernel
                # Catches SystemExit/KeyboardInterrupt too: `sys.exit()` in a
                # cell must end the cell, not the session.
                error = traceback.format_exc()
        finally:
            os.dup2(saved[0], 1)
            os.dup2(saved[1], 2)
            os.close(saved[0])
            os.close(saved[1])
        sink.seek(0)
        raw = sink.read().decode("utf-8", "replace")

    # Anything a child process wrote lands after this cell's own prints.
    text, truncated = _truncate(out.getvalue() + raw)
    return text, result, error, capture_figure(), truncated


def main() -> None:
    # Reserve a private copy of fd 1 for the protocol BEFORE any cell runs,
    # then hand the original fd 1 over to cell capture. Everything the host
    # reads is written to this descriptor and nothing else can reach it.
    protocol_fd = os.dup(1)
    protocol = os.fdopen(protocol_fd, "w", encoding="utf-8", newline="\n")

    # Force UTF-8 on the stdio protocol regardless of the OS locale. On Windows,
    # piped stdin defaults to the ANSI code page (e.g. cp936/GBK), which
    # corrupts non-ASCII source like `print("中文")` before it is executed.
    reconfigure = getattr(sys.stdin, "reconfigure", None)
    if reconfigure is not None:
        reconfigure(encoding="utf-8")

    ns: dict = {"__name__": "__main__"}
    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue
        try:
            req = json.loads(line)
        except json.JSONDecodeError:
            continue
        stdout, result, error, image, truncated = run_cell(ns, req.get("code", ""))
        resp = {
            "id": req.get("id"),
            "ok": error is None,
            "stdout": stdout,
            "result": result,
            "error": error,
            "image": image,
            "truncated": truncated,
        }
        # ensure_ascii keeps the line 7-bit clean, so the host never depends on
        # the pipe's encoding to frame a response.
        protocol.write(json.dumps(resp, ensure_ascii=True) + "\n")
        protocol.flush()


if __name__ == "__main__":
    main()
